src/passes/Monomorphize.cpp

```diff
@@ -47,44 +81,218 @@
+  // regular).
+  std::vector<Expression*> operands;
+
+  // Whether the call is dropped. TODO
+  bool dropped;
+
+  bool operator==(const CallContext& other) const {
+    // We consider logically equivalent expressions as equal (rather than raw
+    // pointers), so that contexts with functionally identical shape are
+    // treated the same.
+    if (operands.size() != other.operands.size()) {
+      return false;
+    }
+    for (Index i = 0; i < operands.size(); i++) {
+      if (!ExpressionAnalyzer::equal(operands[i], other.operands[i])) {
+        return false;
+      }
+    }
+
+    return dropped == other.dropped;
+  }
+
+  bool operator!=(const CallContext& other) const { return !(*this == other); }
+
+  // Build the context from a given call. This builds up the context operands as
+  // as explained in the comments above, and updates the call to send any
+  // remaining values by updating |newOperands| (for example, if all the values
+  // sent are constants, then |newOperands| will end up empty, as we have
+  // nothing left to send).
+  void buildFromCall(Call* call,
+                     std::vector<Expression*>& newOperands,
+                     Module& wasm) {
+    Builder builder(wasm);
+
+    for (auto* operand : call->operands) {
+      // Process the operand. This is a copy operation, as we are trying to move
+      // (copy) code from the callsite into the called function. When we find we
+      // can copy then we do so, and when we cannot that value remains as a
+      // value sent from the call.
+      operands.push_back(ExpressionManipulator::flexibleCopy(
+        operand, wasm, [&](Expression* child) -> Expression* {
+          if (canBeMovedIntoContext(child)) {
+            // This can be moved, great: let the copy happen.
+            return nullptr;
+          }
+
+          // This cannot be moved, so we stop here: this is a value that is sent
+          // into the monomorphized function. It is a new operand in the call,
+          // and in the context operands it is a local.get, that reads that
+          // value.
+          auto paramIndex = newOperands.size();
+          newOperands.push_back(child);
+          // TODO: If one operand is a tee and another a get, we could actually
+          //       reuse the local, effectively showing the monomorphized
+          //       function that the values are the same. (But then the checks
+          //       later down to is<LocalGet> would need to check index too.)
+          return builder.makeLocalGet(paramIndex, child->type);
+        }));
+    }
+
+    // TODO: handle drop
+    dropped = false;
+  }
+
+  // Checks whether an expression can be moved into the context.
+  bool canBeMovedIntoContext(Expression* curr) {
+    // Constant numbers, funcs, strings, etc. can all be copied, so it is ok to
+    // add them to the context.
+    return Properties::isSingleConstantExpression(curr);
```

Comment:
It looks like this is not true for constant `global.get`, but that might be good to include as well.

Reply:
Good point, yeah, this could capture anything "copyable" really. I added a TODO.